File: BOJ/2146/2146.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <cstring>
#define LEN 100
#define MAX 987654321
using namespace std;
typedef pair<int,int> pii;

int N, ans;
int A[LEN + 1][LEN + 1];
bool visited[LEN + 1][LEN + 1];
int dir[4][2] = { {1,0},{-1,0},{0,1},{0,-1} };

void init() {
    cin >> N;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            cin >> A[i][j];
        }
    }
    ans = MAX;
}

void go(int x, int y) {
    if (A[x][y] == 0) return;
    A[x][y] = 0;
    for (auto d : dir) {
        int nx = x + d[0];
        int ny = y + d[1];
        if (nx < 1 || ny < 1 || nx > N || ny > N) continue;
        go(nx, ny);
    }
}

void flood() {
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            if (A[i][j]) {
                memset(visited, false, sizeof(visited));
                queue<pii> q, edge_q;
                q.push({i, j});
                visited[i][j] = true;
                while (!q.empty()) {
                    auto [x,  y] = q.front();
                    q.pop();
                    for (auto & d : dir) {
                        int nx = x + d[0];
                        int ny = y + d[1];
                        if (nx < 1 || ny < 1 || nx > N || ny > N) continue;
                        if (visited[nx][ny]) continue;
                        if (A[nx][ny]) {
                            visited[nx][ny] = true;
                            q.push({nx, ny});
                        }
                        else {
                            edge_q.push({x, y});
                        }
                    }
                }
                int cnt = 0;
                bool flag = true;
//                cout << "edgeq size: " << edge_q.size() << '\n';
                while (!edge_q.empty() && flag) {
                    queue<pii> temp_q;
                    while (!edge_q.empty() && flag) {
                        auto [x, y] = edge_q.front();
//                        if (cnt == 1) cout << x << ", " << y << '\n';
                        edge_q.pop();
                        for (auto & d : dir) {
                            int nx = x + d[0];
                            int ny = y + d[1];
                            if (nx < 1 || ny < 1 || nx > N || ny > N) continue;
                            if (visited[nx][ny]) continue;
                            if (A[nx][ny]) {
                                ans = min(cnt, ans);
                                flag = false;
                                break;
                            }
                            visited[nx][ny] = true;
                            temp_q.push({nx, ny});
                        }
                    }
                    swap(temp_q, edge_q);
                    cnt++;
                }
                go(i, j);
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(0);cin.tie(0);cout.tie(0);
    init();
    flood();
    cout << ans << '\n';
    return 0;
}

File: BOJ/2146/2146.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int N;
    static final int LAND = 1;
    static final int WATER = 0;
    static final int LEN = 100;
    static int[][] A = new int[LEN + 1][LEN + 1];
    static boolean[][] visited = new boolean[LEN + 1][LEN + 1];
    static final int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};

    private static class Point {
        int x;
        int y;
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static int pI(String s) {
        return Integer.parseInt(s);
    }
    static void setVisited() {
        for (int i = 0; i <= LEN; i++) {
            for (int j = 0; j <= LEN; j++) visited[i][j] = false;
        }
    }
    static void init() throws Exception {
        N = pI(br.readLine());
        for (int i = 0; i < N; i++) {
            String[] token = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                A[i][j] = pI(token[j]);
            }
        }
    }
    static void go(int x, int y) {
        if (A[x][y] == WATER) return;
        A[x][y] = WATER;
        for (int[] d : dir) {
            int nx = x + d[0];
            int ny = y + d[1];
            if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
            go(nx, ny);
        }
    }
    static int solve() throws Exception {
        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (A[i][j] == WATER) continue;
                setVisited();
                Queue<Point> q = new LinkedList<>();
                Queue<Point> edgeQ = new LinkedList<>();
                q.offer(new Point(i, j));
                visited[i][j] = true;
                while (!q.isEmpty()) {
                    Point p = q.poll();
                    for (int[] d : dir) {
                        int nx = p.x + d[0];
                        int ny = p.y + d[1];
                        if (nx < 0 || ny < 0 || nx >= N || ny >= N || visited[nx][ny]) continue;
                        if (A[nx][ny] == LAND) {
                            visited[nx][ny] = true;
                            q.offer(new Point(nx, ny));
                        } else {
                            edgeQ.offer(new Point(p.x, p.y));
                        }
                    }
                }
                int cnt = 0;
                boolean flag = true;
                while (!edgeQ.isEmpty() && flag) {
                    Queue<Point> tempQ = new LinkedList<>();
                    while (!edgeQ.isEmpty() && flag) {
                        Point p = edgeQ.poll();
                        for (int[] d : dir) {
                            int nx = p.x + d[0];
                            int ny = p.y + d[1];
                            if (nx < 0 || ny < 0 || nx >= N || ny >= N || visited[nx][ny]) continue;
                            if (A[nx][ny] == LAND) {
                                ans = Math.min(ans, cnt);
                                flag = false;
                                break;
                            }
                            visited[nx][ny] = true;
                            tempQ.offer(new Point(nx, ny));
                        }
                    }
                    edgeQ = tempQ;
                    cnt++;
                }
                go(i, j);
            }
        }
        return ans;
    }
    public static void main(String[] args) throws Exception {
        init();
        bw.write("" + solve());
        bw.flush();
    }
}